viper/parser/stmt.py

```diff
@@ -185,6 +191,83 @@ def parse_for(self):
         self.context.forvars[varname] = True
         return o
 
+    def _is_list_iter(self):
+        """
+        Test if the current statement is a type of list, used in for loops.
+        """
+
+        # Check for literal or memory list.
+        iter_var_type = self.context.vars.get(self.stmt.iter.id).typ if isinstance(self.stmt.iter, ast.Name) else None
+        if isinstance(self.stmt.iter, ast.List) or isinstance(iter_var_type, ListType):
+            return True
+
+        # Check for storage list.
+        if isinstance(self.stmt.iter, ast.Attribute):
+            iter_var_type = self.context.globals.get(self.stmt.iter.attr)
+            if iter_var_type and isinstance(iter_var_type.typ, ListType):
+                return True
+
+        return False
+
+    def parse_for_list(self):
+        from .parser import (
+            parse_body,
+            make_setter
+        )
+
+        iter_list_node = Expr(self.stmt.iter, self.context).lll_node
+        if not isinstance(iter_list_node.typ.subtype, BaseType):  # Sanity check on list subtype.
+            raise StructureException('For loops allowed only on basetype lists.', self.stmt.iter)
+        iter_var_type = self.context.vars.get(self.stmt.iter.id).typ if isinstance(self.stmt.iter, ast.Name) else None
+        subtype = iter_list_node.typ.subtype.typ
+        varname = self.stmt.target.id
+        value_pos = self.context.new_variable(varname, BaseType(subtype))
+        i_pos = self.context.new_variable('_i', BaseType(subtype))
```

Comment:
`_i` is bad form, would probably cause an error in doublely nested lists or something. Either deterministic name-mangling (`indexVar = 'i_' + namehash(varname)` where `namehash` is a hashing function with small output e.g. 8 bytes), or name extension (`indexVar = 'index_' + varname`) should be employed to avoid potential collisions

Reply:
Currently we don't support iterating over nested lists, because we don't allow comparison of lists either, I wrote a specific exception test for it. But will definitely make the indexVar change thanks ;)